compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEInsertDmaBdChain.cpp

```diff
@@ -0,0 +1,276 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/IR/AMDAIEOps.h"
+#include "iree-amd-aie/Transforms/Passes.h"
+#include "iree-amd-aie/Transforms/Utils/AMDAIEDmaUtils.h"
+#include "iree-amd-aie/Transforms/Utils/AMDAIEUtils.h"
+#include "iree-amd-aie/aie_runtime/Utils/ChannelBdIdGenerator.h"
+#include "iree-amd-aie/aie_runtime/iree_aie_runtime.h"
+#include "mlir/IR/Iterators.h"
+#define DEBUG_TYPE "iree-amdaie-insert-dma-bd-chain"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+using TileConnect = std::pair<AMDAIE::TileOp, AMDAIE::ConnectionOp>;
+
+/// Utility function to update `next_bd` and `start_bd` operands.
+LogicalResult updateChainOperands(
+    IRRewriter &rewriter, SmallVector<AMDAIE::NpuHalfDmaCpyNdOp> &dmaChain) {
```

Comment:
```suggestion
    IRRewriter &rewriter, ArrayRef<AMDAIE::NpuHalfDmaCpyNdOp> dmaChain) {
```

Reply:
https://github.com/nod-ai/iree-amd-aie/blob/be58fb4330fbad96abeee154773f14cb9aea1074/compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEInsertDmaBdChain.cpp#L28-L45
A bit sad that the `getBdId` function here is not constant...I don't mind casting here, but maybe `ArrayRef` is not very suitable for any operation type?

Reply:
Oh, good point